File: src/domain/schemas/history.py
"""History-related Pydantic schemas for API.

Provides schemas for viewing and filtering validation history.
"""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field

from src.domain.schemas.validation import FindingResponse
from src.storage.models import ValidationStatus


class HistoryListItem(BaseModel):
    """Summary item for history list view.

    Combines ValidationResult and Document data for display.
    """

    id: UUID = Field(description="ValidationResult ID")
    document_id: UUID = Field(description="Document ID")
    document_filename: str = Field(description="Original document filename")
    status: ValidationStatus = Field(description="Validation result status")
    created_at: datetime = Field(description="When validation was performed")
    findings_count: int | None = Field(
        default=None, description="Number of findings in this validation"
    )

    model_config = {"from_attributes": True}


class HistoryListResponse(BaseModel):
    """Paginated response for validation history list.

    Supports filtering by:
    - status: Filter by validation status (APPROVED, REJECTED, REVIEW_NEEDED)
    - start_date: Filter validations on or after this date
    - end_date: Filter validations on or before this date
    - document_name: Partial match search on document filename (case-insensitive)

    Filters combine with AND logic.
    """

    items: list[HistoryListItem] = Field(description="Validation history items")
    total: int = Field(description="Total number of matching records")
    page: int = Field(ge=1, description="Current page (1-indexed)")
    page_size: int = Field(ge=1, le=100, description="Items per page")
    has_next: bool = Field(description="Whether there are more pages")
    has_prev: bool = Field(description="Whether there are previous pages")


class DocumentInfo(BaseModel):
    """Document metadata for history detail view."""

    id: UUID
    filename: str
    file_hash: str
    file_size_bytes: int
    uploaded_at: datetime

    model_config = {"from_attributes": True}


class HistoryDetailResponse(BaseModel):
    """Full validation detail for history view."""

    id: UUID
    document: DocumentInfo
    status: ValidationStatus
    findings: list[FindingResponse]
    validated_at: datetime  # ValidationResult.created_at
    validator_version: str | None  # rule_version
    model_version: str | None
    processing_time_ms: int | None

    # Audit trail info
    findings_count: int
    has_extraction: bool
    has_analysis: bool

File: frontend/src/services/history.ts
/**
 * History service for AuditEng V2
 */
import api from './api';
import { ValidationFinding } from './documents';

export type ValidationStatus =
  | 'APPROVED'
  | 'REJECTED'
  | 'REVIEW_NEEDED'
  | 'PENDING'
  | 'FAILED';

export interface HistoryListItem {
  id: string;
  document_id: string;
  document_filename: string;
  status: ValidationStatus;
  created_at: string;
  findings_count: number | null;
}

export interface HistoryListResponse {
  items: HistoryListItem[];
  total: number;
  page: number;
  page_size: number;
  has_next: boolean;
  has_prev: boolean;
}

export interface HistoryFilters {
  page?: number;
  page_size?: number;
  status?: ValidationStatus;
  start_date?: string; // YYYY-MM-DD
  end_date?: string; // YYYY-MM-DD
  document_name?: string;
}

export interface DocumentInfo {
  id: string;
  filename: string;
  file_hash: string;
  file_size_bytes: number;
  uploaded_at: string;
}

export interface HistoryDetailResponse {
  id: string;
  document: DocumentInfo;
  status: ValidationStatus;
  findings: ValidationFinding[];
  validated_at: string;
  validator_version: string | null;
  model_version: string | null;
  processing_time_ms: number | null;
  findings_count: number;
  has_extraction: boolean;
  has_analysis: boolean;
}

export const historyService = {
  async list(filters: HistoryFilters = {}): Promise<HistoryListResponse> {
    const params = new URLSearchParams();

    if (filters.page) params.append('page', filters.page.toString());
    if (filters.page_size)
      params.append('page_size', filters.page_size.toString());
    if (filters.status) params.append('status', filters.status);
    if (filters.start_date) params.append('start_date', filters.start_date);
    if (filters.end_date) params.append('end_date', filters.end_date);
    if (filters.document_name)
      params.append('document_name', filters.document_name);

    const response = await api.get<HistoryListResponse>(
      `/history/?${params.toString()}`
    );
    return response.data;
  },

  async getDetail(validationId: string): Promise<HistoryDetailResponse> {
    const response = await api.get<HistoryDetailResponse>(
      `/history/${validationId}`
    );
    return response.data;
  },
};

export default historyService;
